heisenbridge/appservice.py

```diff
@@ -25,29 +29,59 @@ async def load(self):
     async def save(self):
         await self.az.intent.set_account_data("irc", self.config)
 
-    async def create_room(self, name: str, topic: str, invite: List[str]) -> str:
-        resp = await self.az.intent.api.request(
-            Method.POST,
-            Path.v3.createRoom,
-            {
-                "visibility": "private",
-                "name": name,
-                "topic": topic,
-                "invite": invite,
-                "is_direct": False,
-                "power_level_content_override": {
-                    "users_default": 0,
-                    "invite": 100,
-                    "kick": 100,
-                    "redact": 100,
-                    "ban": 100,
-                    "events": {
-                        "m.room.name": 0,
-                        "m.room.avatar": 0,  # these work as long as rooms are private
-                    },
+    async def create_room(self, name: str, topic: str, invite: List[str], restricted: str = None) -> str:
+        req = {
+            "visibility": "private",
+            "name": name,
+            "topic": topic,
+            "invite": invite,
+            "is_direct": False,
+            "power_level_content_override": {
+                "users_default": 0,
+                "invite": 100,
+                "kick": 100,
+                "redact": 100,
+                "ban": 100,
+                "events": {
+                    "m.room.name": 0,
+                    "m.room.avatar": 0,  # these work as long as rooms are private
                 },
             },
-        )
+        }
+
+        if restricted is not None:
+            req["room_version"] = "9"
+            req["initial_state"] = [
+                {
+                    "type": "m.room.join_rules",
+                    "state_key": "",
+                    "content": {
+                        "join_rule": "restricted",
+                        "allow": [{"type": "m.room_membership", "room_id": restricted}],
+                    },
+                }
+            ]
+
+        try:
+            resp = await self.az.intent.api.request(
+                Method.POST,
+                Path.v3.createRoom,
+                req,
+            )
+        except (MUnsupportedRoomVersion, MInvalidRoomState):
+            if restricted is None:
+                raise
+
+            logging.info("Creating restricted room failed.", exc_info=True)
```

Comment:
If we check for the supported room version we don't need to have this code path.